runtime/amd64.S

```diff
@@ -535,6 +535,15 @@ ENDFUNCTION(G(caml_allocN))
 /* Call a C function from OCaml */
 /******************************************************************************/
 
+/* Update [young_limit] when returning from C calls */
+#define RET_FROM_C_CALL                         \
+        cmpq    $0, Caml_state(action_pending); \
+        jne     1f;                             \
+        ret;                                    \
+1:                                              \
+        movq    $-1, Caml_state(young_limit);   \
+        ret
```

Comment:
For a start, adding this to the .S is only one way of implementing it. I looked around for other ways and it seemed the simplest to me, but maybe you know better ways. Notice that we only need it at the end of non-noalloc C calls, so for easier portability it might cheap enough to call `caml_update_young_limit_after_c_call` instead (if there is a good place to insert such a call).

Reply:
1. What happens if you don't update the young limit here? Will the signal be delayed arbitrarily or forgotten? Is it a matter of timely handling of the pending action or is it a matter of correctness?
2. Needs to be implemented for the newer backends. RISC-V, s390x, WIP (Power) @xavierleroy.

Reply:
@xavierleroy An explanation of the new `action_pending` mechanism is given at [signals.c#R281-R305](https://github.com/ocaml/ocaml/pull/11307/files#diff-93ee9a0b40685be35da1909f57cba82cfe51d5a8de495947e5068a8b8d8172cbR281-R305). The explanation is intended to be more complete and self-contained than the text it replaces, thanks to this approach being simpler. Let me know if this text is helpful; if unclear I can try to clarify it, and if still unclear I have a metaphor for 6-year olds (involving a postman and a flag on the mailbox) that I can write down when time allows.

Reply:
@kayceesrk The handling of actions can be delayed until the next minor collection.

Reply:
@kayceesrk It thought about it further—if the next minor collection also happens at an allocation from C, then the callbacks are delayed again until the next minor collection, and so on. So although unlikely, they can in theory be delayed an arbitrarily long time (e.g. code that spends much time on the C side). Also, the minor collection could also fall while another thread is scheduled, which in theory can again delay callbacks an arbitrarily long time (e.g. if the thread that must run an async callback allocates very little compared to the other threads).

Reply:
Thanks for the explanation. 

The code improves on the minor possibility that asynchronous actions may be arbitrarily delayed. This is done by checking that there are pending asynchronous actions when returning from an C call. If so, the young limit is set to a large value so that the next allocation in OCaml fails. 

Benchmarking previously done has shown that this does not add a significant overhead.

Reply:
This, and also the delaying that would happen for sure would be detrimental for memprof callbacks which should run as soon as possible.

Reply:
I would have loved to read the story with the postman and the mailbox :-)   But, yes, the comments in the code clearly describe the design and the implementation.  Thanks!

The only question that popped in my mind (but I think I have an answer) is this: when we're about to return from `caml_c_call` and notice that `action_pending` is set, why don't we jump directly to C code that can handle the pending action, instead of setting `young_limit` and relying on the next poll point in the OCaml code to notice and handle the pending action?

One possible reply is that the frame descriptor that ocamlopt attached to the C call site is not quite right for this purpose: the return value of the C call would have to be kept alive specially, and the backtrace may be wrong if the delayed async action raises an exception.

Another reply is that it would need more code in the runtime/*.S files, something we should avoid for our sanity.

So, I have no further questions on this PR, just one pedantic assembly code remark that can be ignored (next comment).

Reply:
This amounts to adding a polling location when returning from C code. If that's what we want, we can do it, but there is always a danger for code that is careful to control polling locations (especially in a place where that is likely to return a resource). For instance this would break the resource-safety property of `Mutex.protect` https://github.com/ocaml/ocaml/blob/trunk/stdlib/mutex.ml#L27 as it stands, and it would be more work to restore this property.

Reply:
Thanks for the response. The assembly code remark has been fixed in https://github.com/ocaml/ocaml/pull/11307/commits/53dedf18eb79e8f2dbc3edf66e017b2715f7def7. I shall merge this PR.